extensions/wikia/AdEngine/js/recovery/message.js

```diff
@@ -1,44 +1,96 @@
+	'jquery',
 	'wikia.document',
+	'wikia.loader',
+	'wikia.localStorage',
+	'wikia.location',
 	'wikia.log',
+	'wikia.mustache',
 	'wikia.window'
 ], function (
+	adTracker,
 	recoveryHelper,
+	$,
 	doc,
+	loader,
+	localStorage,
+	location,
 	log,
+	mustache,
 	win
 ) {
 	'use strict';
 
 	var logGroup = 'ext.wikia.adEngine.recovery.message',
-		wikiaTopAdsId = 'WikiaTopAds',
-		wikiaRailId = 'WikiaRail';
+		localStorageKey = 'rejectedRecoveredMessage',
+		headerText = 'Hey! It looks like you\'re using ad blocking software!',
+		messageText = 'Wikia runs ads so we can keep the lights on and so Wikia will always be free to use. ' +
+			'We can bring you fun, free, fan-oriented content until my glorious return to Earh if you ' +
+			'<strong>add us to your adblock whitelist</strong>. ' +
+			'<a class="action-accept">Click</a> my face to refresh after you\'re done!';
+
+	function accept() {
+		adTracker.track('recovery/message', 'accept');
+		location.reload();
+	}
+
+	function reject(messageContainer) {
+		adTracker.track('recovery/message', 'reject');
+		messageContainer.style.display = 'none';
+		localStorage.setItem(localStorageKey, true);
+	}
+
+	function isRejected() {
+		return !!localStorage.getItem(localStorageKey);
+	}
+
+	function getTemplate() {
+		return $.when(
+			loader({
+				type: loader.MULTI,
+				resources: {
+					mustache: 'extensions/wikia/AdEngine/templates/recovered_message.mustache'
+				}
+			})
+		).then(function (response) {
+			return response.mustache[0];
+		});
+	}
 
 	function createMessage(uniqueClassName) {
-		var className = 'recovered-message',
-			div = doc.createElement('div');
+		return getTemplate().then(function (template) {
+			var params = {
+					positionClass: 'recovered-message-' + uniqueClassName,
+					header: headerText,
+					text: messageText
+				};
 
-		div.textContent = 'Hello world!';
-		div.classList.add(className, className + '-' + uniqueClassName);
-		return div;
+			return $(mustache.render(template, params));
+		}).then(function (messageContainer) {
+			messageContainer.find('.icon,.action-accept').on('click', accept);
+			messageContainer.find('.close-button').on('click', function () {
+				reject(messageContainer);
+			});
+
+			return messageContainer;
+		});
 	}
 
 	function injectTopMessage() {
-		var topAds = doc.getElementById(wikiaTopAdsId),
-			message = createMessage('top');
-
 		log('recoveredAdsMessage.recover - injecting top message', 'debug', logGroup);
-		topAds.parentNode.insertBefore(message, topAds);
+		createMessage('top').done(function (messageContainer) {
+			$('#WikiaTopAds').before(messageContainer);
+		});
 	}
 
 	function injectRightRailMessage() {
-		var rail = doc.getElementById(wikiaRailId),
-			message = createMessage('right-rail');
-
 		log('recoveredAdsMessage.recover - injecting right rail message', 'debug', logGroup);
-		rail.insertBefore(message, rail.firstChild);
+		createMessage('right-rail').done(function (messageContainer) {
```

Comment:
We have to be careful with promises. And while reading through the changes I was wondering if we should add handling failures everywhere when we use promises or only here. What do you think?